### ISSUE TITLE
Proposal: New `ARG` behavior and `SET`

### ISSUE THREAD

**Issue by vladaionescu** [earthly/earthly#1575]
##### Problem 1: Inefficient cache

* `ARG`s in Earthly differ from `ARG` in Dockerfiles by the fact that Earthly `ARG`s are also env vars
* Because they are also env vars, every `RUN` command automatically depend on the `ARG`s, even if the program(s) being run don't actually use those env vars
* This leads to inefficient use of cache
* This is most painful when there are global `ARG`s defined and their value is changed ever so slightly. Even if only a subset of the targets may be using that global `ARG`, most targets will need to be re-executed.

##### Problem 2: Global `ARG`s have non-obvious behaviors

* When an `ARG` is declared in a base target, it automatically becomes a global arg. This subtle difference in behavior is not obvious to a newcomer. Ideally, an `ARG` declared in a non-base target is the same as an arg declared in the base target.
* Combined with problem 1, this can lead to situations where the cache is used very inefficiently, if the user intends to declare a local `ARG` in base, but inadvertently declares a global (there is currently no way to declare a local `ARG` in base)

##### Problem 3: Redeclaring an ARG with a different value behaves in non-obvious ways

The following snippet does not work and there is no error thrown:

```Earthly
ARG FOO=1
IF ...
    ARG FOO=2
END
```

When an ARG is redeclared it actually does not get its value overridden. The reason is that the value on the right of `=` is actually a new **default value**, not a new true **value**. This can be very surprising for the user.

##### Problem 4: No way to change a value of an `ARG` within the same target

* There is currently no way to change the value of an `ARG` after it has been declared

### Proposal

To address the issues above, we propose the following changes:

Problem 1:

* `ARG`s don’t also set env vars by default
* `ARG --env FOO=bar` can be used to also set the env var (basically brings back the old functionality)
* We replace `ARG`s in `RUN` commands the same way we do for all other commands, and don’t rely on the shell to replace via env vars
* A benefit is that builtin args would not need to be pre-declared, because we could automatically detect where they are used exactly, and therefore we would not cause random cache busts
* The same is true for global `ARG`s - changing a value of a global `ARG` would only influence cache for the commands that actually use that `ARG`
* This all means that `ARG`s would work the same as Dockerfile `ARG`s (helps with better Dockerfile parity)

Problem 2:

* Global ARGs need to be declared as global explicitly:
   ```Earthly
   # Not a global
   ARG FOO
   # Global
   ARG --global FOO
   target:
      # Local
      ARG FOO
      # Results in error
      ARG --global FOO
   ```

Problems 3 & 4: 

* Redeclaring an `ARG` in the same target returns an error, to prevent unintended behavior.
* To change the value of an `ARG`, a new command, `SET` is introduced

```Earthly
# Good
ARG FOO=1
IF …
	SET FOO=2
END
```

```Earthly
# Error if arg not declared in advance
SET FOO=1
```

```Earthly
# Error due to redeclaration
ARG FOO=1
IF …
	ARG FOO=2
END
```

```Earthly
# Using a global within a target requires no redeclaration
ARG --global FOO=1
…
target:
	COPY foo/bar/$FOO ./
```

```Earthly
# Redeclaration of a global is allowed. The inner FOO is a completely different variable in this case
ARG --global FOO=1
…
target:
	ARG FOO=2
```

```Earthly
# Allowed. The inner FOO is a completely different variable, but it takes the value of the global
ARG --global FOO=1
…
target:
	ARG FOO=$FOO
```

```Earthly
# Allowed. The inner FOO is a completely different variable,
# but it takes the value of the global, and within that target
# it also exposes the arg as an env-var
ARG --global FOO=1
…
target:
	ARG --env FOO=$FOO
```

```Earthly
# This will also work, but would not be recommended
# (it can be surprising to the user when value is
# overridden from outside the target)
IF …
	ARG FOO=2
ELSE
	ARG FOO=1
END
```

### Additional (edge) cases

```Earthly
ARG --global FOO=A
target:
   ARG FOO=${FOO}B
   echo "hello $FOO"
```

Execute:
```bash
earthly +target --FOO=C
```

In this case, the output ends up being `hello C`.

### Rollout

This is a breaking change, so it will have to go under a new `VERSION` number.

### Ack

Thanks to @ido-namely and @alexcb for feedback

**Comment by vladaionescu** [comment]
Possible idea re

```earthly
IF …
	ARG FOO=2
ELSE
	ARG FOO=1
END
```

An option could be to not allow `ARG` within `IF` and `FOR` at all and instead to force the user to use `SET` in those cases.

**Comment by rrjjvv** [comment]
> We replace ARGs in RUN commands the same way we do for all other commands, and don’t rely on the shell to replace via env vars

> This all means that ARGs would work the same as Dockerfile ARGs

Would *setting* the `ARG` using the shell (e.g. `ARG foo=$(whoami)`) continue to work after this proposal?  And would that functionality extend to `SET` as well?  (The latter quote, taken literally, suggests it could go away.)

**Comment by vladaionescu** [comment]
> Would setting the ARG using the shell (e.g. ARG foo=$(whoami)) continue to work after this proposal? And would that functionality extend to SET as well? (The latter quote, taken literally, suggests it could go away.)

Yes to both.

**Comment by brandonSc** [comment]
Another suggestion on `ARG` based on community feedback: Exit immediately when an ARG marked as `--required` is missing. This would save users time debugging. 

That suggestion could also belong in a separate ticket.

**Comment by ipizhenko** [comment]
What is implementation statis of this at the moment? Is it already available or just a proposal?

**Comment by danielschlegel** [comment]
In addition an ARG that is just used in a SAVE statement should not invalidate the cache. 

FROM scratch
RUN echo $changing_this_arg_will_invalidate_cache
SAVE IMAGE base-name-$env # having different values for env should not invalidate docker cache.

**Comment by alexcb** [comment]
we should evaluate if a `ARG --unset ...` or `ARG --delete ...` feature will be required.

**Comment by nelsam** [comment]
I think that this statement:

> `ARG`s in Earthly differ from `ARG` in Dockerfiles by the fact that Earthly `ARG`s are also env vars

may be inaccurate. It doesn't change the rest of the points (about cache, especially with respect to global variables), but I _do_ think that `Dockerfile` `ARG`s are also env vars, similar to Earthly.

I've been playing around with `docker` a bit in an attempt to make `earthly` behave like `docker`, and it sure seems like `Dockerfile`s present `ARG`s to the steps of `docker build` as environment variables. The quickest and easiest `Dockerfile` to show that off (credit to @alexcb):

```Dockerfile
FROM alpine:3.15

ARG foo=bar

RUN env | grep foo
```

And the output of `docker build`:

```
[...]
Step 2/3 : ARG foo=bar
 ---> Running in 6b525281550b
Removing intermediate container 6b525281550b
 ---> 337750595f1b
Step 3/3 : RUN env | grep foo
 ---> Running in 74f9dfc1f20a
foo=bar
```

**Comment by mwopfner** [comment]
If I understand the problem correct, the current `ARG` behavior is as follow.
We have an Earthfile and after running it one everything is cached.
```
VERSION 0.7

ARG --global REGISTRY="docker.io"
ARG --global UNUSED="foo"

all:
  FROM alpine
  ARG EARTHLY_VERSION
  ARG PRINT="Hello World"
  ARG NAME="my-image"

  RUN echo "Earthly Version: $EARTHLY_VERSION"
  RUN echo "Output: $PRINT"

  SAVE IMAGE --push $REGISTRY/$NAME
```
If now any of the `ARG`s  (even the `UNUSED` arg) nothing from the cache can be reused, because it is unknown where the ARG is used.

## Desired behavior

- setting arg `UNUSED` should not invalidate the cache for `+all`
- setting arg `NAME` should not invalidate the cache of both `RUN` commands
- setting arg `PRINT` should not invalidate the cache of the first `RUN` command

## Problem
as I understand it.
- Earthly does/can not track where the `ARG` is used.
- The main problem is the `RUN` command because `ARG`s are set as environment variables for the shell call. This includes global `ARG`. Is this also why `Build in ARG` must be specified for a target? Otherwise all targets would always depend on all `Build in ARG`?
- It is currently not possible to track if the `RUN` command uses a specific `ARG` or any `ARG` at all.

## Solutions
These are just some ideas that came to my mind and my help to find a good solution.
1. Specify at the `RUN` command which `ARG`s are used (e.g. ` RUN --ARG=PRINT echo "Output: $PRINT"`). 
2. Add a key-value dictionary for all earthly `ARG`s in the `RUN` command (e.g. ` RUN echo "Output: ${EARTLY_ARG[PRINT]}"`) This should allow to check if a `ARG` is used by some string parsing. It would also be possible to put all `Build in ARG` into the dictionary and get rid of the declaration for the target. 
3. Parse the `RUN` command an understand what it actually does and which variables are used in which scope. This could maybe achieved with the help of https://github.com/mvdan/sh

**Comment by mwopfner** [comment]
@nelsam Thanks for the detailed explanation. I have not thought about the different shells and trying to support all of them is really a mountain of work and would even not solve the problem when a binary, which used in the `RUN`, accesses a variable.

So probably the only robust way, with an appropriate amount of work is to have an extra flag `RUN --explicit-args` with a list of `ARG`s that are available in `RUN`. Thus an easy copy form a Dockerfile still works but for power users it is also possible to get the best use of the cache. Perhaps an additional `RUN --no-env` arg (as proposed before), to mark `RUN` commands that use no `ARG`s at all, would be nice. Or it could just be `--explicit-args` with no args specified.

**Comment by alexcb** [comment]
this has been done with `LET` and `SET`, which was released in `VERSION 0.8`.
